src/Tasks/Delete.cs

```diff
@@ -60,46 +93,60 @@ public void Cancel()
+                    try
                     {
-                        if (FileSystems.Default.FileExists(file.ItemSpec))
+                        // For speed, eliminate duplicates caused by poor targets authoring
+                        if (!deletedFilesSet.Contains(file.ItemSpec))
                         {
-                            // Do not log a fake command line as well, as it's superfluous, and also potentially expensive
-                            Log.LogMessageFromResources(MessageImportance.Normal, "Delete.DeletingFile", file.ItemSpec);
+                            if (FileSystems.Default.FileExists(file.ItemSpec))
+                            {
+                                // Do not log a fake command line as well, as it's superfluous, and also potentially expensive
+                                Log.LogMessageFromResources(MessageImportance.Normal, "Delete.DeletingFile", file.ItemSpec);
+
+                                File.Delete(file.ItemSpec);
+                            }
+                            else
+                            {
+                                Log.LogMessageFromResources(MessageImportance.Low, "Delete.SkippingNonexistentFile", file.ItemSpec);
+                            }
+
+                            // keep a running list of the files that were actually deleted
+                            // note that we include in this list files that did not exist
+                            ITaskItem deletedFile = new TaskItem(file);
+                            deletedFilesList.Add(deletedFile);
+                        }
+                    }
+                    catch (Exception e) when (ExceptionHandling.IsIoRelatedException(e))
+                    {
+                        if (retries < Retries)
+                        {
+                            retries++;
+                            Log.LogWarningWithCodeFromResources("Delete.Retrying", file.ToString(), retries, RetryDelayMilliseconds, e.Message);
 
-                            File.Delete(file.ItemSpec);
+                            Thread.Sleep(RetryDelayMilliseconds);
+                            continue;
                         }
                         else
                         {
-                            Log.LogMessageFromResources(MessageImportance.Low, "Delete.SkippingNonexistentFile", file.ItemSpec);
+                            LogError(file, e);
+                            break;
                         }
-
-                        // keep a running list of the files that were actually deleted
-                        // note that we include in this list files that did not exist
-                        ITaskItem deletedFile = new TaskItem(file);
-                        deletedFilesList.Add(deletedFile);
                     }
+                    // Add even on failure to avoid reattempting
+                    deletedFilesSet.Add(file.ItemSpec);
```

Comment:
I think this should be outside the while loop.

Reply:
Yes, should move this out. Thank you